CONTRIBUTING.md

```diff
@@ -1,11 +1,83 @@
+
+**Examples**
+
+- [#385](https://github.com/getavalon/core/issues/385)
+- [#390](https://github.com/getavalon/core/issues/390)
+- [#415](https://github.com/getavalon/core/issues/415)
+
+<br>
+
+### Bug
+
+When you run into bugs, you can submit an issue. In order for it to be resolved quickly, a bug report should contain:
+
+| # |          | Description
+|--|:---------|:-
+| 1 | **Problem** | What happened? Keep it short
+| 2 | **Reproducible** | How can someone else encounter the bug? Keep it minimal
+| 3 | **Attempts** | What have you tried so far?
+| 4 | **Solution** | Optional. If you've got an idea of what to do about it, but would like to talk about it before diving into code, here's your chance
+
+It's hard to make an exact checklist out of these, as it depends on the particular bug. 
+
+**Examples**
+
+- [#389](https://github.com/getavalon/core/issues/389)
+- [#377](https://github.com/getavalon/core/issues/377)
+- [#412](https://github.com/getavalon/core/issues/412)
+
+<br>
+
+### Pull Request
+
+Here's a few things to keep in mind as you contribute code to Avalon. The overall goal is getting code merged as quickly as possible. Ambiguity cause delays, treat written English as code; it should be clear and concise. 
+
+1. Every new line of code needs purpose and motivation
+1. Every removed line of code needs a reason
+
+**Description**
+
+When you present your PR, here's what you need to do.
+
+1. Briefly summarise the changes, everybody loves bullet lists
+1. Refer to a related issue, if one exists. Else, make one, and refer to [Feature Request](#feature-request) above.
+2. Adhere to code quality standards
+
+**Code Quality**
+
+Here are a few things to keep in mind with regards to code quality.
+
+| # |          | Description
+|--|:---------|:-
+| 1A | **Minimal indirection** | In order to cope with the cognitive overhead of traversing code, keep relevant code together.
+| 1B | **Only separate what is shared** | If a function is only ever used by one module, it should become part of that module. Likewise for modules used by a package.
+| 1C | **Prefer fewer large modules to many small** | Whenever you import a module, you create a dependency. The less of those there are, the better. That means no modules with just a single class.
+| 1D | **Upward and lateral imports** | A module may reach up and sideways in the package hierarchy, but not down. E.g. `maya\lib.py` may reach `avalon\io.py`, but `io.py` may not reach into `maya`.
+| 1E | **Shallow dependency tree** | Avoid traversing more than 3 levels anywhere, unless there is good reason. 3 is plenty.
+| 1F | **Group by dependency, not type** | That is, if 6 modules share code, they should share a package. If 12 functions are used by a single module, they should be part of that module.
+| 2 | **No Subjective Code** | To avoid bikeshedding, avoid submitting code that without a right nor wrong answer. Sometimes, there's a missing space or underscore somewhere
```

Comment:
>  Sometimes, there's a missing space or underscore somewhere

What do you mean with this?

Reply:
.. or missing dot at the end of line. :D

Reply:
Haha, oh no. Yes I was struggling with that one, looks like I forgot to finish.

Not sure how to phrase it; I'd like to avoid PRs that cannot be reviewed. Where opinion A is equal to opinion B. But it's hard to know what's subjective.. any ideas?

Reply:
Is it not just a matter of having a goal with the PR?
If a PR doesn't have a right or wrong answer then it probably doesn't have a goal?

The implementation it self can always be discussed and I'm sure as long as the goal gets archived nobody will that strongly about how its done.

Reply:
Yeah, let's try that.